### ISSUE TITLE
Enable authentication via CILogon

### ISSUE THREAD

**Issue by yuvipanda** [2i2c-org/infrastructure#315]
### Problem statement

Right now we only offer two ways to authenticate for our hubs: GitHub and Google OAuth. However, this falls short in a few ways:

- Many organization-wide hubs will want to authenticate by a given domain (e.g., `@berkeley.edu`) rather than with an individual list of users
- For organizations that don't with to use Google OAuth or GitHub, we do not have a means of authenticating their users.

### Solution

[CILogon](https://www.cilogon.org/home) is an authentication system focused on north american & western european academia. It is the authentication mechanism behind the "eduroam" network and a few others ([some more info is here](https://www.cilogon.org/faq)).

It is free for 'academic users' (https://www.cilogon.org/subscribe), and I would like to think we qualify.

If we support CILogon as a login provider, it lets us provide direct institutional logins to a large swath of potential users. This is probably particularly attractive to communities that are specific to an organization / institution, because it allows them to use the same login credentials that they are already using in their institution.

### Appetite

1 sprint

If we are integrating this with Auth0, it hopefully shouldn't be too much work to get running. It might take a bit of time to get the authentication approval from the CILogon folks, however.

### Guidance / implementation

CILogin has [specific instructions](https://www.cilogon.org/auth0) on integrating with Auth0, and that is what we should do! 

#### 2022-01-26

We'll aim to get the following done by the next sprint:

- [x] Get a valid approved client registered with CILogon (https://cilogon.org/oauth2/register)
- [x] Integrate that with [our Auth0 setup](https://www.cilogon.org/auth0)
- [x] Make it possible to set up CILogon for our hubs: #823
- [x] Choose a hub to use as a "demo" or "sandbox" hub for use by 2i2c team
- [x] Test out CILogon with this hub (maybe a demo hub). Two use-cases to consider:
  - Allow all users that match `@place.edu` pattern
  - Allow a list of users explicitly
  - Allow for a combination of `@place.edu` users *AND* 2i2c admin users
- [x] Use this experience to write up documentation about how to do this.
- [x] Investigate a deeper integration with CILogon to use for other authentication patterns as well :arrow_right: https://github.com/2i2c-org/infrastructure/issues/967
  - Look into their [service offerings](https://www.cilogon.org/subscribe)
  - Look into the [CILogon Authenticator for JupyterHub](https://github.com/jupyterhub/oauthenticator/blob/main/oauthenticator/cilogon.py)
  - Questions to answer:
    - Would this simplify our login process + make it more flexible?
    - Does CILogon have all of the same features that Auth0 does? What are the differences?
    - CILogon costs ~$13,000 a year, is that worth the value we'd get?
    - Is there a "trial run" that we can use to try out their API?
- [ ] Come up with some questions we have for the CILogon team in case we'd like a meeting with them

**Comment by choldgraf** [comment]
After some conversations with @colliand I think that we should bump this up the priority list a little bit, because it will be really useful if we can tell universities that we will "probably" be able to integrate with their institutional SSO via CILogon.

**Comment by GeorgianaElena** [comment]
Some updates about this:

I followed the instructions in https://www.cilogon.org/auth0  and created a new auth0 application that I then registered with CILogon. I provided the  2i2c `support` email address in the form (hope that's ok).

We need to wait for this to happen now:
*An administrator will contact you once your registration request has been approved.*

However, even if the request hasn't been approved, we've been provided the client credentials. Should we store these encrypted somewhere in this repo (maybe [here](https://github.com/2i2c-org/infrastructure/blob/master/config/secrets.yaml))?

**Comment by consideRatio** [comment]
@GeorgianaElena it seems the request was approved, see https://2i2c.freshdesk.com/a/tickets/38

**Comment by GeorgianaElena** [comment]
Great! Thanks @consideRatio!

**Comment by damianavila** [comment]
> Should we store these encrypted somewhere in this repo (maybe here)?

That location makes sense to me from the top of my head.

**Comment by choldgraf** [comment]
Hey @GeorgianaElena - just checking in to see if there's anything that others can do to help you out or unblock on this one!

**Comment by GeorgianaElena** [comment]
I realized that we're supporting JupyterHub OAuthenticator now, so I'm experimenting with using JupyterHub OAuthenticator to enable CILogon instead of using it through Auth0. I think this issue was opened long before the OAuthenticator addition, so the Auth0 integration might not necessarily be a requirement.

Happy to hear people's ideas about this.

Thank you for asking, @choldgraf!

**Comment by GeorgianaElena** [comment]
## Authentication via CILogon status

### 1. Why not use JupyterHub [`CILogonOAuthenticatior`](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/cilogon.py#L45)?

Now that we support authentication via JupyterHub Native OAuthenticator, why not use [`CILogonOAuthenticatior`](https://github.com/jupyterhub/oauthenticator/blob/master/oauthenticator/cilogon.py#L45) directly, without going through Auth0?

**Advantages**
- CILogonOAuthenticatior natively supports restricting access based on a list of allowed idps
- One CILogon client for all the hubs (similar with Auth0), without adding the Auth0 layer of complexity
- We don't have to maintain or worry about the CILogon social connection [added manually](https://www.cilogon.org/auth0) to our 2i2c account

**Negatives**
- When registering the CILogon client, I had to provide the list of allowed callback URLs. All *allowed_callback_urls* that we intend to use need to be in this list and need to be known at the time of client registration. Otherwise, any changes to this list can only happen through  an email sent to CILogon (and we cannot know these in advance because they are based on hub URLs). 

### 2. Integrate CILogon with our Auth0 setup

This has been successfull and https://github.com/2i2c-org/infrastructure/pull/823 adds support for automating enabling this connection from our config files.
We only needed to list the 2i2c auth0 domain in the callback_url list during CILogon registration, and after CILogon authentication, Auth0 knows how to redirect back to each hub's address.

**HOWEVER**
- Restricting users based on idp will be ugly:
  - Mills hub username pattern example: https://github.com/2i2c-org/infrastructure/blob/d3da0f3e722bec6b3a5bc72a5614d0c7f6bbfd08/config/hubs/cloudbank.cluster.yaml#L367

* Another approach would be to make changes upstream, either in `JupyterHub Authenticator` or directly in `Auth0Oauthenticator`(though I'm not sure if it would make sense to have an allowed email domain list in this Auth0 specific authenticator ?)

* Auth0 rules?? (we are only allowed 3 rules I believe, so this is not scalable)

### 3. Help

I would love help and feedback about which of these two approaches is better, especially:
1. Is there a way to have only the 2i2c.cloud domain in the [callback_url list of CILogon](https://cilogon.org/oauth2/register) and achieve the same redirection behaviour like we do with the 2i2c auth0 domain?
2. If 1'st point :arrow_up: is not achievable, what's the best way to "de-uglify" the `username_pattern` example above and restrict based on identity provider?

**Comment by sgibson91** [comment]
> * When registering the CILogon client, I had to provide the list of allowed callback URLs. All _allowed_callback_urls_ that we intend to use need to be in this list and need to be known at the time of client registration. Otherwise, any changes to this list can only happen through  an email sent to CILogon (and we cannot know these in advance because they are based on hub URLs).

To me, this is a big blocker. I'd rather cope with ugly code than need to rely on manual intervention from CILogon

**Comment by choldgraf** [comment]
So it sounds like our next step here is to:

- Go with @GeorgianaElena's proposed #2 option above in https://github.com/2i2c-org/infrastructure/issues/315#issuecomment-975887945 
- Accept that we will have an ugly `username_pattern` value for CILogon, and try to resolve this in some future issue

Is that right?

**Comment by choldgraf** [comment]
I've updated the top comment with a few more next steps that I think are needed to close this out. I also added a few scenarios that we should test to make sure it works as expected!

**Comment by choldgraf** [comment]
### How to allow community users + 2i2c users to log-in

- JupyterHub only allows you to authenticate with a single kind of account (e.g. EITHER GitHub OR gmail)
- If we could use an auth provider that supports multiple *kinds of logins* then we could still use one auth provider, but that *provider* could authenticate with different email addresses.
- CILogon *might* be able to allow for multiple login sources. Could we use a university address + a github account?
- Will it be possible to authenticate ourselves if we are using a SSO?
- Could we create a token for an "admin user" and use this instead of the typical sign-in process?
  - This feels like it'd be a security concern, it'd be a "root access" token :-/
- Could we use CILogon *instead* of Auth0, and just use this to authenticate across all of our 

### Next steps

- [ ] Add documentation about our CILogon setup, so that we can try it out
- [ ] Learn more about whether CILogon could provide the multiple authentication pathway

**Comment by consideRatio** [comment]
@GeorgianaElena as a followup to a support email received in https://2i2c.freshdesk.com/a/tickets/69, I wanted to note that:

Terry, a CILogon administrator, suggests that it would be possible to use a paid CILogin API to automate things, which is what I suspect could be directly related to addressing the negative of using CILogonOAuthenticator you've described.

> **Negatives**
> 
> * When registering the CILogon client, I had to provide the list of allowed callback URLs. All _allowed_callback_urls_ that we intend to use need to be in this list and need to be known at the time of client registration. Otherwise, any changes to this list can only happen through  an email sent to CILogon (and we cannot know these in advance because they are based on hub URLs).

I'll respond to Terry that I've forwarded this idea.

**Comment by choldgraf** [comment]
### Update: Try out the demo hub!

@GeorgianaElena has deployed a JupyterHub at demo.pilot.2i2c.cloud - this can be used to try out things, demo them, etc. She's enabled CILogon with `@2i2c.org` as an authenticated domain. Give it a shot!

My quick feedback: I really like it! It seems like a nice way to support many different institutional sign-ons, and I really like the simplicity of setup!

**Comment by GeorgianaElena** [comment]
> Question: how do I "log out" of CILogon?

Good question @choldgraf, I hadn't noticed. I believe there's something similar with this issue we had some time ago
https://github.com/2i2c-org/infrastructure/issues/422. For CILogon I think we need to also to logout the users out of [Identity Provider Session Layer](https://auth0.com/docs/authenticate/login/logout).

I also noticed a similar behavior with the toronto hub :confused: (That one is using [AzureAdOAuthenticator](https://github.com/jupyterhub/oauthenticator/blob/main/oauthenticator/azuread.py)) I will open up issues about each of these issues.

**Comment by GeorgianaElena** [comment]
### Update

I opened https://github.com/2i2c-org/infrastructure/issues/967 to tackle the remaining bullets of this issue (about changing current Auth0 setup with CILogon) and https://github.com/2i2c-org/infrastructure/issues/957 about the logging out issue that @choldgraf noticed.

I was thinking to add the logging out issue to out Operational list of tasks and discuss more about changing Auth0 with CILogon in its own issue, linked above. WDYT @choldgraf? for me, it feels like since we have CILogon auth, we can close this one?

**Comment by choldgraf** [comment]
@GeorgianaElena that sounds good to me - my main question is whether we can provide a workaround for people who run into the issue described in https://github.com/2i2c-org/infrastructure/issues/957 - I think it'd be a big problem if users logged in with the wrong email but couldn't log out. Could we recommend that they clear their browser cache or something?

Regarding https://github.com/2i2c-org/infrastructure/issues/967, agreed that we can follow that up in subsequent conversations.

**Comment by GeorgianaElena** [comment]
Note for the future :arrow_right: 
Possible workarounds for #957 could be https://github.com/2i2c-org/infrastructure/pull/974 and https://github.com/2i2c-org/pilot-homepage/pull/8

**Comment by choldgraf** [comment]
OK, now that #974 is in, I think that we can consider this one resolved! 🎉🎉🎉

There are still some things to keep track of and follow up. Specifically:

- As people start using this, we'll likely need to make focused improvements to our docs or implementation, but we can follow up on that as-needed.
- I'll reach out to the CILogon team to ask them about the potential for partnering with their service. @GeorgianaElena has an issue to track that here: #967
